geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java

```diff
@@ -0,0 +1,236 @@
+  private static MemberVM server3;
+  private static int[] redisPorts;
+  private static Jedis jedis;
+
+  private SessionRepository<Session> sessionRepository;
+  private ConfigurableApplicationContext springContext;
+
+  @BeforeClass
+  public static void classSetup() {
+    Properties locatorProperties = new Properties();
+    locatorProperties.setProperty(MAX_WAIT_TIME_RECONNECT, "15000");
+
+    locator = cluster.startLocatorVM(0, locatorProperties);
+
+    server1 = startRedisVM(1, 0);
+    server2 = startRedisVM(2, 0);
+    server3 = startRedisVM(3, 0);
+
+    redisPorts = new int[] {
+        cluster.getRedisPort(1),
+        cluster.getRedisPort(2),
+        cluster.getRedisPort(3)};
+
+    jedis = new Jedis("localhost", redisPorts[0], JEDIS_TIMEOUT);
+  }
+
+  private static MemberVM startRedisVM(int vmId, Integer redisPort) {
+    int locatorPort = locator.getPort();
+
+    return cluster.startRedisVM(vmId, x -> x
+        .withProperty(REDIS_PORT, redisPort.toString())
+        .withConnectionToLocator(locatorPort));
+  }
+
+  @Before
+  @SuppressWarnings("unchecked")
+  public void setup() {
+    String[] args = new String[] {
+        "" + redisPorts[2],
+        "" + redisPorts[1],
+        "" + redisPorts[0]};
+
+    springContext = SpringApplication.run(RedisSpringTestApplication.class, args);
+    sessionRepository = springContext.getBean(SessionRepository.class);
+    assertThat(sessionRepository).isNotNull();
+  }
+
+  @After
+  public void teardown() {
+    springContext.stop();
+    jedis.flushAll();
+    sessionIds.clear();
+  }
+
+  @Test
+  public void sessionOperationsDoNotFail_whileServersAreRestarted() throws Exception {
+    createSessions();
+
+    AtomicBoolean running = new AtomicBoolean(true);
+    AtomicReference<String> phase = new AtomicReference<>("STARTUP");
+
+    Future<Integer> future1 = executor.submit(() -> sessionUpdater(0, running, phase));
+    Future<Integer> future2 = executor.submit(() -> sessionUpdater(1, running, phase));
+
+    GeodeAwaitility.await().during(1, TimeUnit.SECONDS).until(() -> true);
+
+    phase.set("CRASH 1 SERVER2");
+    cluster.crashVM(2);
+    server2 = startRedisVM(2, redisPorts[1]);
+    phase.set("CRASH 1 REBALANCING");
+    rebalanceAllRegions(server2);
+
+    phase.set("CRASH 2 SERVER3");
+    cluster.crashVM(3);
+    server3 = startRedisVM(3, redisPorts[2]);
+    phase.set("CRASH 2 REBALANCING");
+    rebalanceAllRegions(server3);
+
+    phase.set("CRASH 3 SERVER2");
+    cluster.crashVM(2);
+    server2 = startRedisVM(2, redisPorts[1]);
+    phase.set("CRASH 3 REBALANCING");
+    rebalanceAllRegions(server2);
+
+    phase.set("CRASH 4 SERVER3");
+    cluster.crashVM(3);
+    server3 = startRedisVM(3, redisPorts[2]);
+    phase.set("CRASH 4 REBALANCING");
+    rebalanceAllRegions(server3);
+
+    phase.set("FINISHING");
+    GeodeAwaitility.await().during(10, TimeUnit.SECONDS).until(() -> true);
+
+    running.set(false);
+    future1.get();
+    future2.get();
+  }
```

Comment:
Could you validate that the values on the session are correct, or at least not garbage? 
This could theoretically pass if the session updater threads were all stalled for the ten seconds the test is running for. We should confirm that all the attributes we expect attr-#-[0->NUM_SESSIONS] exist.
Even better would be to validate that the count at the end of the value is greater than or equal to the modCount at the end of the attribute name. This would indicate that we made it through more than one iteration and no data was lost in the rebalancing.

Reply:
Added some validation.